sail/src/main/java/org/apache/rya/rdftriplestore/inference/InferenceEngine.java

```diff
@@ -723,29 +724,38 @@ public void handleStatement(final Statement statement) throws RDFHandlerExceptio
                     }
                 }
             }
-
-            final List<Statement> typeStatements = new ArrayList<>();
-            ryaDaoQueryWrapper.queryAll(type, OWL.INTERSECTIONOF, null, new RDFHandlerBase() {
-                @Override
-                public void handleStatement(final Statement statement) throws RDFHandlerException {
-                    typeStatements.add(statement);
-                }
-            });
+            for (final Set<Resource> intersection : intersectionList) {
+                addIntersection(intersection, type);
+            }
+        }
+        for (final Entry<Resource, List<Set<Resource>>> entry : intersectionsProp.entrySet()) {
+            final Resource type = entry.getKey();
+            final List<Set<Resource>> intersectionList = entry.getValue();
 
             final Set<URI> superClasses = getSuperClasses((URI) type);
+            for (final URI superClass : superClasses) {
+                // Add intersections to super classes if applicable.
+                // IF:
+                // :A intersectionOf[:B, :C]
+                // AND
+                // :A subclassOf :D
+                // Then we can infer:
+                // intersectionOf[:B, :C] subclassOf :D
+                for (final Set<Resource> intersection : intersectionList) {
+                    addIntersection(intersection, superClass);
+                }
+            }
+            // Check if other keys have any of the same intersections and infer
+            // the same subclass logic to them that we know from the current
+            // type. Propagating up through all the superclasses.
             for (final Set<Resource> intersection : intersectionList) {
-                addIntersection(intersection, type);
-                for (final URI superClass : superClasses) {
-                    // Add intersections to super classes if applicable.
-                    // IF:
-                    // :A intersectionOf[:B, :C]
-                    // AND
-                    // :A subclassOf :D
-                    // Then we can infer:
-                    // intersectionOf[:B, :C] subclassOf :D
-                    for (final Statement statement : typeStatements) {
-                        final Resource intersectionOfBnode = (Resource) statement.getObject();
-                        addSubClassOf(intersectionOfBnode, superClass);
+                final Set<Resource> otherKeys = Sets.newHashSet(intersectionsProp.keySet());
+                otherKeys.remove(type);
+                for (final Resource otherKey : otherKeys) {
+                    if (intersectionsProp.get(otherKey).contains(intersection)) {
+                        for (final URI superClass : superClasses) {
```

Comment:
I updated this. otherKey and type are now inferred to be equivalentClasses so this changed some of the unit test expected output.